benchmark/spmv/spmv.cpp

```diff
@@ -0,0 +1,385 @@
+    std::shared_ptr<const gko::Executor> exec, const gko::matrix_data<> &data)
+{
+    auto mat = MatrixType::create(std::move(exec));
+    mat->read(data);
+    return mat;
+}
+
+
+const std::map<std::string, std::function<std::unique_ptr<gko::LinOp>(
+                                std::shared_ptr<const gko::Executor>,
+                                const gko::matrix_data<> &)>>
+    matrix_factory{{"csr", read_matrix<gko::matrix::Csr<>>},
+                   {"coo", read_matrix<gko::matrix::Coo<>>},
+                   {"ell", read_matrix<gko::matrix::Ell<>>},
+                   {"hybrid", read_matrix<gko::matrix::Hybrid<>>},
+                   {"sellp", read_matrix<gko::matrix::Sellp<>>}};
+
+// executor mapping
+const std::map<std::string, std::function<std::shared_ptr<gko::Executor>()>>
+    executor_factory{
+        {"reference", [] { return gko::ReferenceExecutor::create(); }},
+        {"omp", [] { return gko::OmpExecutor::create(); }},
+        {"cuda", [] {
+             return gko::CudaExecutor::create(FLAGS_device_id,
+                                              gko::OmpExecutor::create());
+         }}};
+
+
+template <typename RandomEngine>
+std::unique_ptr<vector> create_rhs(std::shared_ptr<const gko::Executor> exec,
+                                   RandomEngine &engine, gko::dim<2> dimension)
+{
+    auto rhs = vector::create(exec);
+    rhs->read(gko::matrix_data<>(
+        dimension, std::uniform_real_distribution<>(-1.0, 1.0), engine));
+    return rhs;
+}
+
+
+template <typename RandomEngine, typename Allocator>
+void spmv_system(const char *format_name,
+                 std::shared_ptr<const gko::Executor> exec,
+                 const gko::matrix_data<> &data, const vector *b,
+                 const vector *x, const unsigned int warm_iter,
+                 const unsigned int run_iter, rapidjson::Value &test_case,
+                 Allocator &allocator, RandomEngine &rhs_engine) try {
+    if (!FLAGS_overwrite && test_case.HasMember(format_name)) {
+        return;
+    }
+
+    add_or_set_member(test_case, format_name,
+                      rapidjson::Value(rapidjson::kObjectType), allocator);
+    auto system_matrix = share(matrix_factory.at(format_name)(exec, data));
+
+    // warm run
+    for (unsigned i = 0; i < warm_iter; i++) {
+        auto x_clone = clone(x);
+        exec->synchronize();
+        system_matrix->apply(lend(b), lend(x_clone));
+        exec->synchronize();
+    }
+    duration_type time(0);
+    // timed run
+    for (unsigned i = 0; i < run_iter; i++) {
+        auto x_clone = clone(x);
+        exec->synchronize();
+        auto tic = std::chrono::system_clock::now();
+        system_matrix->apply(lend(b), lend(x_clone));
+
+        exec->synchronize();
+        auto toc = std::chrono::system_clock::now();
+        time += std::chrono::duration_cast<duration_type>(toc - tic);
+    }
+    add_or_set_member(test_case[format_name], "time(ns)",
```

Comment:
maybe remove `ns`, and just have `time`. [JSONata](http://docs.jsonata.org/) (the thing we use to transform and plot the data) may have trouble with this.

Reply:
removed ns